src/mlpack/core/data/save_image_impl.hpp

```diff
@@ -14,154 +15,94 @@
+          ImageOptions& opts)
 
-  return SaveImage(filename, tmpMatrix, info, fatal);
+{
+  return SaveImage(files, matrix, opts);
 }
 
-// Image saving API for multiple files.
 template<typename eT>
-bool Save(const std::vector<std::string>& files,
-          arma::Mat<eT>& matrix,
-          ImageInfo& info,
-          const bool fatal)
+bool SaveImage(const std::vector<std::string>& files,
+               const arma::Mat<eT>& matrix,
+               ImageOptions& opts)
 {
-  if (files.size() == 0)
+  if (files.empty())
   {
-    if (fatal)
-    {
-      Log::Fatal << "Save(): vector of image files is empty; nothing to save."
-          << std::endl;
-    }
-    else
+    std::stringstream oss;
+    oss << "Save(): vector of image files is empty; nothing to save.";
+    HandleError(oss, opts);
+  }
+
+  // Check if we do have any type that is not supported.
+  for  (size_t i = 0; i < files.size() ; ++i)
+  {
+    if (!opts.saveType.count(Extension(files.at(i))))
     {
-      Log::Warn << "Save(): vector of image files is empty; nothing to save."
-          << std::endl;
+      std::stringstream oss;
+      oss << "Save(): file type " << opts.FileTypeToString()
+          << " isn't supported. Currently image saving supports: ";
+      for (const auto& x : opts.saveType)
+        oss << "  " << x;
+      oss << "." << std::endl;
+      HandleError(oss, opts);
     }
-
-    return false;
   }
 
-  arma::Mat<unsigned char> img;
-  bool status = true;
-
-  for (size_t i = 0; i < files.size() ; ++i)
+  size_t dimension = opts.Width() * opts.Height() * opts.Channels();
+  // We only need to check the rows since it is a matrix.
+  if (dimension != matrix.n_rows)
   {
-    arma::Mat<eT> colImg(matrix.colptr(i), matrix.n_rows, 1,
-        false, true);
-    status &= Save(files[i], colImg, info, fatal);
+    std::stringstream oss;
+    oss << "data::Save(): The given image dimensions do not match the "
+        << "dimensions of the matrix to be saved!";
+    HandleError(oss, opts);
   }
-
-  return status;
-}
-
-inline bool SaveImage(const std::string& filename,
-                      arma::Mat<unsigned char>& image,
-                      ImageInfo& info,
-                      const bool fatal)
-{
-  // Check to see if the file type is supported.
-  if (!ImageFormatSupported(filename, true))
+  // Unfortunately we cannot move because matrix is const.
+  arma::Mat<eT> tempMatrix;
+  MakeAlias(tempMatrix, matrix, matrix.n_rows, matrix.n_cols, 0, true);
+  bool success = false;
+  for (size_t i = 0; i < files.size() ; ++i)
   {
-    std::ostringstream oss;
-    oss << "Save(): file type " << Extension(filename) << " not supported.\n";
-    oss << "Currently image saving supports ";
-    for (auto extension : SaveFileTypes())
-      oss << "  " << extension;
-    oss << "." << std::endl;
-
-    if (fatal)
+    // Update opts.Format() at each iteration.
+    DetectFromExtension<arma::Mat<eT>, ImageOptions>(files.at(i), opts);
```

Comment:
I think this only needs to be done if the format is `Image` or `AutoDetect` though.  However, maybe I misunderstood: does `DetectFromExtension` do nothing if the format is a specific format (like `CSV` or `PNG`)?  If `DetectFromExtension` only does something if the format is `AutoDetect` or `Image`, then you can ignore this comment.